xdsl/dialects/csl.py

```diff
@@ -0,0 +1,250 @@
+            "arg_attrs": arg_attrs,
+            "res_attrs": res_attrs,
+        }
+        super().__init__(properties=properties, regions=[region])
+
+    def verify_(self) -> None:
+        # If this is an empty region (external function), then return
+        if len(self.body.blocks) == 0:
+            return
+
+        entry_block: Block = self.body.blocks[0]
+        block_arg_types = [arg.type for arg in entry_block.args]
+        if self.function_type.inputs.data != tuple(block_arg_types):
+            raise VerifyException(
+                "Expected entry block arguments to have the same types as the function "
+                "input types"
+            )
+
+    @classmethod
+    def parse(cls, parser: Parser) -> FuncOp:
+        (
+            name,
+            input_types,
+            return_types,
+            region,
+            extra_attrs,
+            arg_attrs,
+        ) = parse_func_op_like(
+            parser, reserved_attr_names=("sym_name", "function_type", "sym_visibility")
+        )
+
+        assert len(return_types) <= 1, "csl.func can't have more than one result type!"
+
+        func = cls(
+            name=name,
+            function_type=(input_types, return_types[0] if return_types else None),
+            region=region,
+            arg_attrs=arg_attrs,
+        )
+        if extra_attrs is not None:
+            func.attributes |= extra_attrs.data
+        return func
+
+    def print(self, printer: Printer):
+        print_func_op_like(
+            printer,
+            self.sym_name,
+            self.function_type,
+            self.body,
+            self.attributes,
+            arg_attrs=self.arg_attrs,
+            reserved_attr_names=(
+                "sym_name",
+                "function_type",
+                "sym_visibility",
+                "arg_attrs",
+            ),
+        )
+
+
+@irdl_op_definition
+class ReturnOp(IRDLOperation):
+    """
+    Return for CSL operations such as functions and tasks.
+    """
+
+    name = "csl.return"
+
+    ret_val = opt_operand_def(Attribute)
+
+    traits = frozenset([HasParent(FuncOp), IsTerminator()])
+
+    def __init__(self, return_val: SSAValue | Operation | None = None):
+        super().__init__(operands=[return_val])
+
+    def verify_(self) -> None:
+        func_op = self.parent_op()
+        assert isinstance(func_op, FuncOp)
+
+        if tuple(func_op.function_type.outputs) != tuple(
+            val.type for val in self.operands
+        ):
+            raise VerifyException(
+                "Expected arguments to have the same types as the function output types"
+            )
+
+    def print(self, printer: Printer):
+        print_return_op_like(printer, self.attributes, self.operands)
+
+    @classmethod
+    def parse(cls, parser: Parser) -> ReturnOp:
+        attrs, args = parse_return_op_like(parser)
+        op = ReturnOp(*args)
+        op.attributes.update(attrs)
+        return op
```

Comment:
Sidenote: this gave me the push to do a bugfix #2475 and assuming this fix, this can use declarative syntax instead.